creds/vault/cache.go

```diff
@@ -0,0 +1,116 @@
+package vault
+
+import (
+	"sync"
+	"time"
+
+	vaultapi "github.com/hashicorp/vault/api"
+)
+
+type cachedSecret struct {
+	deadline time.Time
+	secret   *vaultapi.Secret
+}
+
+// A Cache caches secrets read from a SecretReader until the lease on
+// the secret expires. Once expired the credential is proactively
+// deleted from cache to maintain a smaller cache footprint.
+type Cache struct {
+	sync.RWMutex
+	cache    map[string]*cachedSecret
+	newItems chan time.Time
+	sr       SecretReader
+}
+
+// TODO: Should a cache have a max size to
+// prevent unbounded growth?
```

Comment:
if we do that, does it then rather make sense going for redis in a term, that this is already offered there and its also a memory cache and flushed when we restart the services ( security reasons ).